config/FindXenomai.cmake

```diff
@@ -80,10 +80,38 @@ if( XENOMAI_CFLAGS_ERROR)
     message(FATAL_ERROR "Could not determine cflags with command ${XENOMAI_XENO_CONFIG} --skin=${XENOMAI_SKIN_NAME} --cflags ${XENO_CONFIG_LDFLAGS_EXTRA_ARGS}")
 endif()
 
-string(REGEX MATCHALL "-L([^ ]+)|-l([^ ]+)|.*bootstrap(-pic)?.o" XENOMAI_LIBRARY ${XENOMAI_LDFLAGS})
-string(REGEX MATCHALL "-I([^ ]+)" XENOMAI_INCLUDE_DIR ${XENOMAI_CFLAGS})
-string(REGEX MATCHALL "-D([^ ]+)" XENOMAI_COMPILE_DEFINITIONS ${XENOMAI_CFLAGS})
-string(REPLACE "-I" ";" XENOMAI_INCLUDE_DIR ${XENOMAI_INCLUDE_DIR})
+string(STRIP XENOMAI_LDFLAGS ${XENOMAI_LDFLAGS})
+string(STRIP XENOMAI_CFLAGS ${XENOMAI_CFLAGS})
+string(REPLACE " " ";" _XENOMAI_LDFLAGS ${XENOMAI_LDFLAGS})
+string(REPLACE " " ";" _XENOMAI_CFLAGS ${XENOMAI_CFLAGS})
+set(XENOMAI_LIBRARY)
+set(XENOMAI_INCLUDE_DIR)
+set(XENOMAI_DEFINITIONS)
+set(XENOMAI_OTHER_LDFLAGS)
+set(XENOMAI_OTHER_CFLAGS)
+foreach(_entry ${_XENOMAI_LDFLAGS})
+  string(REGEX MATCH "^-L(.+)|^-l(.+)|^(-Wl,.+)|^(.*bootstrap(-pic)?.o)" _lib ${_entry})
+  if(_lib)
+    list(APPEND XENOMAI_LIBRARY ${_lib})
+  else()
+    list(APPEND XENOMAI_OTHER_LDFLAGS ${_entry})
+  endif()
+endforeach()
+foreach(_entry ${_XENOMAI_CFLAGS})
+  string(REGEX MATCH "^-I.+" _include_dir ${_entry})
+  string(REGEX MATCH "^-D.+" _definition ${_entry})
+  if(_include_dir)
+    string(REGEX REPLACE "^-I" "" _include_dir ${_include_dir})
+    list(APPEND XENOMAI_INCLUDE_DIR ${_include_dir})
+  elseif(_definition)
+    string(REGEX REPLACE "^-D" "" _definition ${_definition})
+    list(APPEND XENOMAI_DEFINITIONS ${_definition})
+  else()
+    list(APPEND XENOMAI_OTHER_CFLAGS ${_entry})
+  endif()
+endforeach()
+string(REPLACE ";" " " XENOMAI_OTHER_LDFLAGS "${XENOMAI_OTHER_LDFLAGS}")
+string(REPLACE ";" " " XENOMAI_OTHER_CFLAGS "${XENOMAI_OTHER_CFLAGS}")
```

Comment:
@meyerj I'm curious about these `XENOMAI_OTHER_CFLAGS` and `XENOMAI_OTHER_LDFLAGS` that are not used elsewhere.

Reply:
At least for Xenomai 2.6 they should be empty because everything is covered by one of the regular expressions above. Could add a check and emit a fatal error if some unexpected arguments are passed here.

I think I added these two for consistency with the [FindPkgConfig](https://cmake.org/cmake/help/v3.0/module/FindPkgConfig.html) cmake helper, but some linker flags which for pkg-config would end up in `<XPREFIX>_LIBRARIES` are added to `XENOMAI_LIBRARY` instead (and can be passed to `target_link_libraries()`). I also just noticed that the suffix is not the same (`_OTHER_CFLAGS` vs `_CFLAGS_OTHER`).

Feel free to add a check or remove it.

Output of `xeno-config` on my Ubuntu 16.04 system with `libxenomai-dev` version 2.6.4+dfsg-0.2:
```sh
johannes@im-laptop-007:/tmp$ xeno-config --skin native --cflags
-I/usr/include/xenomai -D_GNU_SOURCE -D_REENTRANT -D__XENO__
johannes@im-laptop-007:/tmp$ xeno-config --skin native --ldflags
-lnative -L/usr/lib/x86_64-linux-gnu -lxenomai -lpthread -lrt
```

Reply:
Ok but then on Xenomai 3 we have (`xeno-config --cflags --alchemy`): 
```
-I/home/hoarau/dev/xenomai-3.0.8/install/include/cobalt
-I/home/hoarau/dev/xenomai-3.0.8/install/include 
-D_GNU_SOURCE -D_REENTRANT
-fasynchronous-unwind-tables
```
The `-fasynchronous-unwind-tables` will endup in the `_OTHER_CFLAGS` but it not used anywhere. Is that bad ?

Reply:
I just remembered [that line](https://github.com/kuka-isir/rtt_gazebo_robot_interface/blob/master/CMakeLists.txt#L34).
It can definitely be added to XENOMAI_DEFINITIONS

Reply:
Not sure why exactly is is needed by the alchemy skin. This is the commit that introduced it with some explanation: https://www.xenomai.org/pipermail/xenomai-git/2017-November/008972.html

In CMake version >3.0 additional compile flags should be added with [add_compile_options()](https://cmake.org/cmake/help/v3.0/command/add_compile_options.html). But [add_definitions()](https://cmake.org/cmake/help/v3.0/command/add_definitions.html#command:add_definitions) can also be used for flags other than preprocessor definitions and is backwards compatible:
> This command can be used to add any flags, but it is intended to add preprocessor definitions.

So yes, adding it to `XENOMAI_DEFINITIONS` instead will probably work. Or directly use `XENOMAI_OTHER_CFLAGS` in the calling [check_depend.cmake](https://github.com/meyerj/rtt/blob/59646430316e77fb1996ff2cac5b1ff4de7c9448/config/check_depend.cmake) script [here](https://github.com/meyerj/rtt/blob/59646430316e77fb1996ff2cac5b1ff4de7c9448/config/check_depend.cmake#L172).